### ISSUE TITLE
Disk space not freed after syft command

### ISSUE THREAD

**Issue by rmkanda** [anchore/syft#416]
**What happened**:
Whenever I ran syft command to analyse a image the disk usage goes up and once the syft is completed. 
The utilized disk space is not freed.

What is the temporary directory? Can the disk space freed automatically?

**Environment**:
- Output of `syft version`: 0.15.2
- OS MacOS

**Comment by luhring** [comment]
Hi @rmkanda! Thanks for reporting this. Our expectation is that all files we create **_are cleaned up_** after the run completes. (The exceptions to this are in cases where the app crashes or when the app is terminated early.) So this is definitely a problem, and we need to look more into it.

>What is the temporary directory?

We use your OS's tmp directory (seen by running `echo $TMPDIR`), and within that directory, we create a few directories as needed for the analysis job. These directory names should include the names of our projects, like `syft` and `stereoscope`.

I think I'm able to reproduce what you're describing...

### How to reproduce

I have a split screen terminal. On the left, I’m running `watch -n 0.2 'ls -l $TMPDIR | grep -i stereoscope | wc -l'` . On the right, I’m running `syft registry:anchore/anchore-engine:latest` repeatedly. And then I’m watching the count of stereoscope temp dirs go up and down.

I’m finding that this count returns to the starting point _only sometimes_.

I'll be looking more into what might be causing this.

**Comment by DatGameh** [comment]
Hi @luhring, I have a few question about these temp files:

For context, I intend to use Syft for large-volume scans of about ~1000-2000 images at a time, but due to this bug, these large `stereoscope` directories of ~10GB each are created in the `tmp` folder but never deleted. This quickly consumes hundreds of GB of storage in a short period.

![image](https://github.com/anchore/syft/assets/71962500/741a641c-b703-476f-8205-ec9009f3de26)

There are two things I would like to ask regarding this:
1. Can I safely delete all these `stereoscope` and `syft` directories in `tmp` without breaking Syft?
2. Is it possible to define where Syft saves these temp files? 
    * For instance, instead of the OS's 'tmp' directory, save them in a `temp` directory in my project that gets deleted after the program finishes.

Thank you!

**Comment by kzantow** [comment]
@DatGameh first of all, I should note that Syft _should_ be cleaning this stuff up and if it's not, that's a bug we need to fix (I thought it had been, when this was closed). To answer your specific questions:
1. After Syft runs, you can definitely delete those directories, they won't be used again
2. It is possible by using a standard(?) environment variable, such as `TMPDIR`, e.g.: `TMPDIR=$(pwd)/tmp syft python:latest` seems to work just fine.

However, in figuring this out, I definitely am able to reproduce some leftovers that should be cleaned up but are not. For example:
```
$ mdkir tmp
$ TMPDIR=$(pwd)/tmp syft python:latest --from registry
...
$ ls tmp
stereoscope-2691232634
```
Although in this case, the directory has some subdirectories but ultimately they are all empty, so I'm not sure steps offhand to reproduce large content being left on the filesystem. Regardless, these temp directories should get cleaned up, so I'm reopening this issue for that purpose.

**Comment by DatGameh** [comment]
@kzantow Thank you for looking into the issue for me!

Here are extra details which hopefully helps:

The OS I am using is either RHEL 8.9 or Ubuntu 22.04

The task I am doing requires me to scan hundreds of images in a short period.
I do this through a Python script that runs multiple Syft Scans in parallel.
I run up to 30 scans simultaneously.

I used this as reference to implement limited concurrency: https://death.andgravity.com/limit-concurrency#asyncio-semaphore

Here's an example of my script:

```
import asyncio

CONCURRENT_SCANS = 30
images = [...]  # (A list of about 200-300 image names/links)


async def syft_scanner(image):
    scan_process = await asyncio.create_subprocess_shell(f'syft {image} -o spdx-json={image}.spdx.json',
                                                         stdout=asyncio.subprocess.PIPE,
                                                         stderr=asyncio.subprocess.PIPE)

    return scan_process.returncode


def scan_task_generator(images):
    for image in images:
        yield asyncio.create_task(
            syft_scanner(image),
            name=f'syft_scan_{image}'
        )


async def main():
    ## START
    scan_gen = scan_task_generator(images)
    scan_gen_end = False
    pending = set()

    successful_scans = 0
    failed_scans = 0

    while pending or not scan_gen_end:
        while len(pending) < CONCURRENT_SCANS and not scan_gen_end:
            try:
                aw = next(scan_gen)
            except StopIteration:
                scan_gen_end = True
            else:
                pending.add(asyncio.ensure_future(aw))

        if not pending:
            return

        done, pending = await asyncio.wait(
            pending, return_when=asyncio.FIRST_COMPLETED
        )
        while done:
            try:
                done_result = (done.pop()).result()
            except Exception as e:
                failed_scans += 1
                continue
            if done_result == 0:
                successful_scans += 1
            else:
                failed_scans += 1

    print('successful scans: ' + str(successful_scans))
    print('failed scans: ' + str(failed_scans))

    ## END


if __name__ == '__main__':
    asyncio.run(main())

```

Please let me know if you need any additional info. Thanks!

**Comment by kzantow** [comment]
@DatGameh sorry for the delay getting back to you. I've spent quite a bit of time trying to reproduce this but I'm just really not able to. I'm not sure if I possibly killed a process and left some files not cleaned up when I saw this originally, but I can't come up with a way to reproduce it -- even using concurrent scans with a modified version of your provided script. The only thing I can think that's causing this for you is _something_ is killing the process before it terminates normally (including things like out of memory), but without being able to make some sort of steps to reproduce the issue, it's nearly impossible to debug and fix. Is there anything else you could provide to help with reproduction -- maybe a script including a bunch of specific images that exhibits the problem?

**Comment by DatGameh** [comment]
@kzantow 

Thank you for looking into the problem!
As of now, I've been able to avoid the issue entirely by creating a temporary Syft directory that gets deleted whenever my script exits - whether if it ends normally or abruptly stops through a signal, exception or anything else.

Something like this:
![image](https://github.com/anchore/syft/assets/71962500/fb4834ef-6a00-429d-b055-ae02febc2c0b)

There _are_ times when I need to abruptly end the program, since the scanning process naturally takes a long time to complete. In my attempts to recreating the problem so far, prematurely ending my program (e.g. SIGINT) does reproduce the undeleted temp files rather consistently. I didn't think to mention it initially, because I thought temp files should naturally be deleted even if the process terminates prematurely. I apologize for not mentioning it earlier.

I wonder then: Is it expected that ending the process prematurely means the temp files do not get deleted?

So far, I have not been able to reproduce the problem if I allow the program to finish naturally, but I will continue trying to recreate it and report to you if I am able to reproduce it.

**Comment by kzantow** [comment]
Thanks @DatGameh. Given there really hasn't been a way to reproduce this that anyone has found (other than killing the process, which is expected to not execute clean up code), I'm going to close this again. Sorry for the noise here!

If we _do_ find a way to reproduce this, let's open a new issue with the exact steps/scripts/etc. that can be used. Thanks in advance!

**Comment by ttc0419** [comment]
Hi, just want to add that the problem seems still exist. We developed a service with grype go package for our CI pipeline and there are many residual files with old create dates:
```
drwx------ 3 root root 4.0K Nov  4 11:36 stereoscope-1101923093
drwx------ 3 root root 4.0K Oct 24 14:30 stereoscope-1117472616
drwx------ 3 root root 4.0K Oct 30 00:07 stereoscope-1184311838
drwx------ 3 root root 4.0K Oct 22 19:43 stereoscope-1297812466
drwx------ 3 root root 4.0K Oct 30 11:30 stereoscope-1322959633
drwx------ 3 root root 4.0K Nov 13 18:09 stereoscope-1437274622
drwx------ 3 root root 4.0K Nov  9 21:40 stereoscope-1439098261
drwx------ 3 root root 4.0K Nov  6 17:51 stereoscope-1457646508
drwx------ 3 root root 4.0K Nov  5 19:27 stereoscope-1469701708
drwx------ 3 root root 4.0K Nov 11 10:18 stereoscope-1499317168
drwx------ 3 root root 4.0K Nov 11 17:15 stereoscope-1516242438
drwx------ 3 root root 4.0K Oct 28 19:19 stereoscope-1537616066
drwx------ 3 root root 4.0K Nov  7 20:47 stereoscope-1543760877
drwx------ 3 root root 4.0K Oct 25 10:18 stereoscope-1561756302
drwx------ 3 root root 4.0K Oct 31 16:47 stereoscope-1604307189
drwx------ 3 root root 4.0K Oct 29 10:27 stereoscope-1632447167
drwx------ 3 root root 4.0K Nov  7 21:24 stereoscope-1637310209
drwx------ 3 root root 4.0K Nov 11 20:40 stereoscope-1653763920
drwx------ 3 root root 4.0K Oct 22 17:21 stereoscope-1721151727
drwx------ 3 root root 4.0K Oct 28 18:19 stereoscope-176273481
drwx------ 3 root root 4.0K Nov  7 12:46 stereoscope-1792272766
drwx------ 3 root root 4.0K Nov  6 15:01 stereoscope-1848798695
drwx------ 3 root root 4.0K Oct 26 11:36 stereoscope-1886563690
drwx------ 3 root root 4.0K Nov 13 21:06 stereoscope-1907317533
drwx------ 3 root root 4.0K Oct 31 12:15 stereoscope-1916930452
drwx------ 3 root root 4.0K Nov  9 21:39 stereoscope-1917823424
drwx------ 3 root root 4.0K Nov 11 13:12 stereoscope-1929997564
drwx------ 3 root root 4.0K Oct 22 19:22 stereoscope-1954951372
drwx------ 3 root root 4.0K Oct 21 14:47 stereoscope-1957006226
drwx------ 3 root root 4.0K Nov  4 20:32 stereoscope-1974947042
drwx------ 3 root root 4.0K Oct 28 11:35 stereoscope-1985458681
drwx------ 3 root root 4.0K Nov  8 13:11 stereoscope-2073644107
drwx------ 3 root root 4.0K Oct 28 21:37 stereoscope-2096710620
drwx------ 3 root root 4.0K Oct 28 00:33 stereoscope-2099041566
drwx------ 3 root root 4.0K Oct 30 20:32 stereoscope-2130017124
drwx------ 3 root root 4.0K Nov  8 13:35 stereoscope-2136862611
drwx------ 3 root root 4.0K Nov  6 20:21 stereoscope-216651663
drwx------ 3 root root 4.0K Oct 31 09:13 stereoscope-2167876863
drwx------ 3 root root 4.0K Nov  5 19:03 stereoscope-2175131411
drwx------ 3 root root 4.0K Nov 12 10:11 stereoscope-2251595389
drwx------ 3 root root 4.0K Nov  8 11:48 stereoscope-2258405577
drwx------ 3 root root 4.0K Oct 21 19:28 stereoscope-2374300709
drwx------ 3 root root 4.0K Oct 18 16:53 stereoscope-2420384978
drwx------ 3 root root 4.0K Nov  7 00:59 stereoscope-248478620
drwx------ 3 root root 4.0K Nov 11 22:26 stereoscope-2549084931
drwx------ 3 root root 4.0K Nov  5 12:07 stereoscope-270205658
drwx------ 3 root root 4.0K Oct 20 00:37 stereoscope-2709319917
drwx------ 3 root root 4.0K Nov  1 18:57 stereoscope-2832765497
drwx------ 3 root root 4.0K Oct 30 11:36 stereoscope-2886258460
drwx------ 3 root root 4.0K Oct 25 19:52 stereoscope-2940729885
drwx------ 3 root root 4.0K Nov 12 12:46 stereoscope-2968401088
drwx------ 3 root root 4.0K Nov  1 13:28 stereoscope-298529153
drwx------ 3 root root 4.0K Oct 22 08:08 stereoscope-3046498098
drwx------ 3 root root 4.0K Oct 29 14:18 stereoscope-3064097362
drwx------ 3 root root 4.0K Nov 13 20:19 stereoscope-3122723547
drwx------ 3 root root 4.0K Nov 14 14:45 stereoscope-3139436098
drwx------ 3 root root 4.0K Nov  6 14:54 stereoscope-3167620550
drwx------ 3 root root 4.0K Oct 28 19:02 stereoscope-324264017
drwx------ 3 root root 4.0K Oct 30 09:47 stereoscope-3296139294
drwx------ 3 root root 4.0K Oct 29 11:06 stereoscope-3299908870
drwx------ 3 root root 4.0K Nov  7 13:28 stereoscope-336810944
drwx------ 3 root root 4.0K Nov  6 19:12 stereoscope-3374549340
drwx------ 3 root root 4.0K Oct 24 16:00 stereoscope-3375847167
drwx------ 3 root root 4.0K Nov  7 09:33 stereoscope-3409156747
drwx------ 3 root root 4.0K Oct 29 16:16 stereoscope-34105001
drwx------ 3 root root 4.0K Nov  8 17:37 stereoscope-3466400401
drwx------ 3 root root 4.0K Nov 12 23:48 stereoscope-3493087446
drwx------ 3 root root 4.0K Oct 29 18:43 stereoscope-3519304805
drwx------ 3 root root 4.0K Nov  7 20:41 stereoscope-3597332693
drwx------ 3 root root 4.0K Nov 12 12:24 stereoscope-3777192269
drwx------ 3 root root 4.0K Nov  8 15:05 stereoscope-3840777056
drwx------ 3 root root 4.0K Nov 13 15:17 stereoscope-4005356340
drwx------ 3 root root 4.0K Nov  4 15:37 stereoscope-402248866
drwx------ 3 root root 4.0K Nov  4 17:22 stereoscope-4023878299
drwx------ 3 root root 4.0K Nov 13 00:30 stereoscope-4026664397
drwx------ 3 root root 4.0K Oct 24 17:18 stereoscope-4040335007
drwx------ 3 root root 4.0K Oct 28 11:44 stereoscope-4060283570
drwx------ 3 root root 4.0K Oct 24 17:43 stereoscope-407115937
drwx------ 3 root root 4.0K Nov  9 21:39 stereoscope-4077342908
drwx------ 3 root root 4.0K Oct 25 11:48 stereoscope-4172872676
drwx------ 3 root root 4.0K Nov  1 12:23 stereoscope-4293180317
drwx------ 3 root root 4.0K Nov  8 20:02 stereoscope-456715961
drwx------ 3 root root 4.0K Oct 22 14:09 stereoscope-472173829
drwx------ 3 root root 4.0K Nov  6 19:54 stereoscope-540799591
drwx------ 3 root root 4.0K Nov  6 14:36 stereoscope-630864814
drwx------ 3 root root 4.0K Nov  5 16:40 stereoscope-653535124
drwx------ 3 root root 4.0K Nov 12 11:58 stereoscope-718177542
drwx------ 3 root root 4.0K Nov  7 22:26 stereoscope-735312082
drwx------ 3 root root 4.0K Nov  4 09:15 stereoscope-775186007
drwx------ 3 root root 4.0K Nov  9 21:39 stereoscope-789078194
drwx------ 3 root root 4.0K Oct 25 18:00 stereoscope-805387930
drwx------ 3 root root 4.0K Nov  7 12:43 stereoscope-886941102
```

Are there any new developments on this?
